core/shared/src/main/scala/it/unibo/scafi/lib/Gradients.scala

```diff
@@ -5,26 +5,269 @@
 
 package it.unibo.scafi.lib
 
+import scala.concurrent.duration.FiniteDuration
+import scala.math.Numeric.DoubleIsFractional
+import it.unibo.utils.Filters.expFilter
+
 trait StdLib_Gradients {
   self: StandardLibrary.Subcomponent =>
 
   type Metric = ()=>Double
 
+  import Builtins.Bounded
+
+  implicit val idBounded: Bounded[ID]
+
   trait Gradients {
-    self: FieldCalculusSyntax with StandardSensors =>
+    self: FieldCalculusSyntax with StandardSensors with GenericUtils with StateManagement with BlockG =>
 
-    case class Gradient(algorithm: (Boolean,()=>Double)=>Double, source: Boolean=false, metric: Metric = nbrRange) {
+    case class Gradient(algorithm: (Boolean, () => Double) => Double, source: Boolean = false, metric: Metric = nbrRange) {
       def from(s: Boolean): Gradient = this.copy(source = s)
+
       def withMetric(m: Metric): Gradient = this.copy(metric = m)
+
       def run(): Double = algorithm(source, metric)
     }
 
-    val ClassicGradient = Gradient(classicGradient(_,_), false, nbrRange)
+    val ClassicGradient: Gradient = Gradient(classicGradient, source = false, nbrRange)
 
     def classicGradient(source: Boolean, metric: () => Double = nbrRange): Double =
-      rep(Double.PositiveInfinity){ case d =>
-        mux(source){ 0.0 }{ minHoodPlus(nbr(d) + metric()) }
+      rep(Double.PositiveInfinity) { case d =>
+        mux(source) {
+          0.0
+        } {
+          minHoodPlus(nbr(d) + metric())
+        }
       }
-  }
 
+    def hopGradient(source: Boolean): Double =
+      rep(Double.PositiveInfinity) {
+        hops => {
+          mux(source) {
+            0.0
+          } {
+            1 + minHood(nbr {
+              hops
+            })
```

Comment:
Does this formatting come from the IDE? Consider avoiding such formatting to prevent too many lines.. e.g. by using  `nbr(hops)` to make it non-blocky.